sdk/src/groth16/ceremony/phase1.rs

```diff
@@ -0,0 +1,106 @@
+use {
+    crate::groth16::{curve::PairingEngine, error::SetupError},
+    ark_serialize::{CanonicalDeserialize, Compress, Validate},
+    std::{
+        fs::File,
+        io::{BufReader, Read},
+        path::Path,
+    },
+};
+
+/// Phase-1 universal SRS: [tau^i] in G1 and G2.
+/// We require at least `n_g1` and `n_g2` powers respectively.
+pub struct Phase1<E: PairingEngine> {
```

Comment:
Can you try to reduce some of the complexity across the board? It seems that `E` is always `PairingEngine` - do we need this generic? I think this can be simplified in a lot of place. I'm having a hard time to being to understand the code to be fair and this might help a lot.